Sources/OktaAnalytics/OktaAnalytics.swift

```diff
@@ -102,12 +109,199 @@ public final class OktaAnalytics: NSObject {
+            assert(false, "Failed to fetch scenarios")
+        }
+        Self.coreDataStack.managedContext.saveContext()
+
+        let scenarioPropertiesFetchRequest = NSFetchRequest<ScenarioProperty>(entityName: "ScenarioProperty")
+        scenarioPropertiesFetchRequest.predicate = NSPredicate(format: "scenarioID CONTAINS %@", scenarioID)
+        do {
+            let scenarioProperties = try Self.coreDataStack.managedContext.fetch(scenarioPropertiesFetchRequest)
+            for scenarioProperty in scenarioProperties {
+                Self.coreDataStack.managedContext.delete(scenarioProperty)
+            }
+        } catch {
+            assert(false, "Failed to fetch scenario properties")
+        }
+        Self.coreDataStack.managedContext.saveContext()
+    }
+
+    /**
+        end sceanrio with scenario name. send all properties to provider and clears local storage.
+
+        - Parameters:
+           - name: scenario name
+        */
+    public static func endScenario(_ name: Name) {
+        lock.readLock()
+        defer { lock.unlock() }
+
+        func fetchProperties(_ scenarioName: Name) -> [String: String] {
+            var properties = [String: String]()
+            let scenarioPropertiesFetchRequest = NSFetchRequest<ScenarioProperty>(entityName: "ScenarioProperty")
+            scenarioPropertiesFetchRequest.predicate = NSPredicate(format: "name CONTAINS %@", name)
+            do {
+                let scenarioProperties = try Self.coreDataStack.managedContext.fetch(scenarioPropertiesFetchRequest)
+                for scenarioProperty in scenarioProperties {
+                    if let key = scenarioProperty.key {
+                        properties[key] = scenarioProperty.value
+                    }
+                    Self.coreDataStack.managedContext.delete(scenarioProperty)
+                }
+            } catch {
+                assert(false, "Failed to fetch scenario properties")
+            }
+            Self.coreDataStack.managedContext.saveContext()
+            return properties
+        }
+
+        let scenarioFetchRequest = NSFetchRequest<Scenario>(entityName: "Scenario")
+        scenarioFetchRequest.predicate = NSPredicate(format: "name CONTAINS %@", name)
+        do {
+            let scenarios = try Self.coreDataStack.managedContext.fetch(scenarioFetchRequest)
+            for scenario in scenarios {
+                if let displayName = scenario.displayName {
+                    Self.trackEvent(displayName, withProperties: fetchProperties(name))
+                }
+                Self.coreDataStack.managedContext.delete(scenario)
+            }
+        } catch {
+            assert(false, "Failed to fetch scenarios")
+        }
+        Self.coreDataStack.managedContext.saveContext()
+    }
+
+    /**
+        end sceanrio with scenario name. clears local storage.
+
+        - Parameters:
+           - name: scenario name
+        */
+    public static func disposeScenario(_ name: Name) {
```

Comment:
what does this public interface do, why would an app like to dispose  Scenario without sending it

Reply:
it's a use case if client think there are interrupted cases in the db